### PR TITLE
Add load/store functions

### PR DESCRIPTION
**Pull request by calebzulawski**
This came up in zulip, so figured I'd add them.

### CODE REVIEW

**Comment by miguelraz** [comment]
Needs tests!

File: crates/core_simd/src/first.rs [hunk 1]
```diff
@@ -42,6 +47,44 @@ macro_rules! impl_vector {
...
+            pub unsafe fn store_to_slice_unchecked(self, dst: &mut [$type]) {
+                (dst.as_mut_ptr() as *mut Self).write_unaligned(self)
+            }
+
+            /// Stores a vector to a slice.
+            ///
+            /// # Panics
+            /// This function will panic if the length of `dst` is not at least `LANES`.
+            #[inline]
+            pub fn store_to_slice(self, dst: &mut [$type]) {
```
Comment (workingjubilee) [inline crates/core_simd/src/first.rs:84]: I am not sure these names are the way to go: we don't use "load" and "store", we use "read" and "write". This risks being _especially_ silly if it optimizes to something other than `STR`/`ST1`/etc.

Comment (workingjubilee) [inline crates/core_simd/src/first.rs:84]: I am not sure these names are the way to go: we don't use "load" and "store", we use "read" and "write". This risks being _especially_ silly if it optimizes to something other than `STR`/`ST1`/etc.

File: crates/core_simd/src/first.rs [hunk 2]
```diff
@@ -42,6 +47,44 @@ macro_rules! impl_vector {
...
+
+            /// Loads a vector from a slice.
+            ///
+            /// # Panics
+            /// This function will panic if the length of `src` is not at least `LANES`.
+            #[inline]
+            pub fn load_from_slice(src: &[$type]) -> Self {
+                assert!(src.len() >= LANES, "slice must be at least as long as vector");
+                unsafe { Self::load_from_slice_unchecked(src) }
+            }
```
Comment (workingjubilee) [inline crates/core_simd/src/first.rs:68]: I'm fine with offering an unaligned read as the "default" interface, but panic machinery can often fail to optimize out (and even result in double-retests... a clear compiler error, but one we should be aware of). So my question is, do we want to have the infallible version offer a panicking interface or one that masks and zero-fills?

**Review by workingjubilee — COMMENTED** [review]
Missing all those `#[inline]`s was a mistake, but I am not sure the rest of this PR is the way to go. If you want to split those out, we can merge them independently.

File: crates/core_simd/src/first.rs (hunk 1 above)
Comment (calebzulawski) [reply crates/core_simd/src/first.rs:84]: I was on the fence about the names.  At first I was thinking I wanted to distinguish between pointer and slice access, but now I remember that the `Read` and `Write` traits use slices...  On the other hand, SIMD instruction sets seem to use "load" and "store" relatively universally.  Read and write are probably most consistent.

File: crates/core_simd/src/first.rs (hunk 2 above)
Comment (calebzulawski) [reply crates/core_simd/src/first.rs:68]: I think panicking is appropriate.  This function will probably rarely be optimal but it's probably necessary to provide and it's no less efficient than zero-filling, I think.  Either way it's just a branch.  In my opinion `from_array` is the infallible version.

**Comment by calebzulawski** [comment]
I've been thinking a little more about this PR and perhaps the correct way to read from a slice is to obtain an array from the slice (using whatever functions are available--if they aren't they should be added to `std`).  @workingjubilee made a good point somewhere in zulip that an array basically is a vector, the semantics are different, but `from_array` should just be a move instruction at worst.

I still feel like it's possible there's something we should be adding to help reading vectors from slices, but maybe it's more about arrays and not so much vectors.  Should we close this?

**Comment by workingjubilee** [comment]
Yeah, I think we should revisit this later with a different approach when we are more confident there's a special desirable here. We should already be emitting `vmovups` and friends.

**Pull request closed without merging**